google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -42,51 +42,71 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
 }
 
 void TracingMessageBatch::AddMessageSpanMetadata() {
+  auto context = batch_sink_parent_span_->GetContext();
+  auto trace_id = internal::ToString(context.trace_id());
+  auto span_id = internal::ToString(context.span_id());
   for (auto& message_span : message_spans_) {
     message_span->AddEvent("gl-cpp.batch_flushed");
+    message_span->SetAttribute("pubsub.batch_sink.trace_id", trace_id);
+    message_span->SetAttribute("pubsub.batch_sink.span_id", span_id);
   }
 }
 
-void TracingMessageBatch::Flush() {
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> MakeParentSpan(
+    std::vector<
+        opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>> const&
+        message_spans) {
   using opentelemetry::trace::SpanContext;
   using AttributesList =
       std::vector<std::pair<opentelemetry::nostd::string_view,
                             opentelemetry::common::AttributeValue>>;
   auto constexpr kMaxOtelLinks = 128;
   std::vector<std::pair<SpanContext, AttributesList>> links;
-  auto batch_size = message_spans_.size();
+  auto batch_size = message_spans.size();
   links.reserve(batch_size);
-
   // If the batch size is less than the max size, add the links to a single
   // span.
   if (batch_size < kMaxOtelLinks) {
     std::transform(
-        message_spans_.begin(), message_spans_.end(), std::back_inserter(links),
+        message_spans.begin(), message_spans.end(), std::back_inserter(links),
         [i = static_cast<std::int64_t>(0)](auto const& span) mutable {
           return std::make_pair(
               span->GetContext(),
               AttributesList{{"messaging.pubsub.message.link", i++}});
         });
   }
-  auto batch_sink_span_parent =
+  auto batch_sink_parent_span =
       internal::MakeSpan("BatchSink::AsyncPublish",
                          /*attributes=*/
                          {{"messaging.pubsub.num_messages_in_batch",
                            static_cast<std::int64_t>(batch_size)}},
                          /*links*/ links);
+  return batch_sink_parent_span;
+}
 
-  // TODO(#12528): Handle batches larger than 128.
+void TracingMessageBatch::Flush() {
+  decltype(message_spans_) message_spans;
+  {
+    std::lock_guard<std::mutex> lk(mu_);
+    message_spans.swap(message_spans_);
+  }
 
-  // This must be called before we clear the message spans.
-  AddMessageSpanMetadata();
-  // Clear message spans.
-  message_spans_.clear();
+  auto batch_sink_parent_span = MakeParentSpan(message_spans);
 
-  batch_sink_spans_.push_back(batch_sink_span_parent);
+  // TODO(#12528): Handle batches larger than 128.
+  // TODO(alevenb): Refactor AddMessageSpanMetadata to accept message_spans and
+  // batch_sink_parent_span as params instead of using member variables. This
+  // must be called before we clear the message spans.
+  batch_sink_parent_span_ = batch_sink_parent_span;  // Remove this line.
+  AddMessageSpanMetadata();
```

Comment:
I know this does not do what this PR intends anymore, since it is invalidating message_spans_ at the top, but I think putting the refactor in a separate PR will be cleaner. Let me know if you disagree and I can refactor in this PR.

Reply:
Outdated.